### PR TITLE
Issue #289: added support for BigNumeric

### PR DESCRIPTION
**Pull request by himanshukohli09**

### CODE REVIEW

File: python-files/big_numeric_support.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from pyspark.sql.types import UserDefinedType, StringType
```
Comment (davidrabinowitz) [inline python-files/big_numeric_support.py:1]: Please move to the python package of `google.cloud.spark.bigquery`

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
1. Please rename `python-files` to `pythonlib`
2. Please add an acceptance test using BigNumeric, so we can have it running on Dataproc
3. Please fix the Avro write

File: published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestContext.java [hunk 2]
```diff
@@ -21,12 +21,16 @@
...
   final String testBaseGcsDir;
+  final String testBigNumericDataset;
+  final String testBigNumericTable;
 
   public AcceptanceTestContext(String testId, String clusterId) {
     this.testId = testId;
     this.clusterId = clusterId;
     this.testBaseGcsDir = AcceptanceTestUtils.createTestBaseGcsDir(testId);
     this.connectorJarUri = this.testBaseGcsDir + "/connector.jar";
+    this.testBigNumericDataset = "bignumeric_acceptance_test_dataset_" + testId.replace("-", "_");
```
Comment (davidrabinowitz) [inline published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestContext.java:32]: Please add some random string such as uuid

File: published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestUtils.java [hunk 3]
```diff
@@ -125,12 +137,65 @@ protected static void deleteGcsDir(String testBaseGcsDir) throws Exception {
...
                     .spliterator(),
                 false)
             .map(Blob::getBlobId)
             .toArray(BlobId[]::new);
     storage.delete(blobIds);
   }
+
+  public static void createBqDataset(String dataset) {
+    DatasetId datasetId = DatasetId.of(dataset);
+    BigQueryOptions.getDefaultInstance().getService().create(DatasetInfo.of(datasetId));
```
Comment (davidrabinowitz) [inline published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestUtils.java:153]: Please create a private static member `bigquery` which is initialized once

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
What's in the zip file? Why do we need it? It is best to keep simple files - txt, csv, etc.

File: python-files/big_numeric_support.py (hunk 1 above)
Comment (himanshukohli09) [reply python-files/big_numeric_support.py]: done

File: published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestUtils.java [hunk 4]
```diff
@@ -27,14 +33,18 @@
...
 import java.util.Comparator;
 import java.util.stream.StreamSupport;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
 
 public class AcceptanceTestUtils {
 
   // must be set in order to run the acceptance test
-  private static final String BUCKET = System.getenv("ACCEPTANCE_TEST_BUCKET");
+  private static final String BUCKET = "kohli_desc";//System.getenv("ACCEPTANCE_TEST_BUCKET");
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestUtils.java]: removed

File: published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestContext.java (hunk 2 above)
Comment (himanshukohli09) [reply published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestContext.java:32]: done

**Comment by himanshukohli09** [comment]
> 2\. g

Done

**Comment by himanshukohli09** [comment]
> What's in the zip file? Why do we need it? It is best to keep simple files - txt, csv, etc.

we need __init__.py for creating the python modules and hence we need the zip file

File: published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestUtils.java (hunk 3 above)
Comment (himanshukohli09) [reply published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestUtils.java:153]: Cannot initialise the dataset once, as every acceptance calls its own "setup" and "tearDown".

Comment (davidrabinowitz) [reply published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestUtils.java:153]: The `bigquery` member should be initialized once.

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
The standard for python package is with the leading `com.` (unlike Java). Please remove this package.

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
please rebase on master, build.sbt has merging conflicts

**Comment by himanshukohli09** [comment]
> please rebase on master, build.sbt has merging conflicts

Done

**Comment by himanshukohli09** [comment]
> The standard for python package is with the leading `com.` (unlike Java). Please remove this package.

Done

File: published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestUtils.java (hunk 3 above)
Comment (himanshukohli09) [reply published/src/test/java/com/google/spark/bigquery/acceptance/AcceptanceTestUtils.java]: done

File: README.md [hunk 5]
```diff
@@ -619,6 +619,79 @@ the field's description which includes the spark type of the field.
...
+val rows: Array[java.math.BigDecimal] = df
+  .collect()
+  .map(row => row.get({columnPositionOfBignumeric}).asInstanceOf[BigNumeric].number)
+
+rows.foreach(value => System.out.println("BigNumeric value  " + value.toPlainString))
+```
+
+**Python:** For spark's [UserDefinedType](https://spark.apache.org/docs/1.4.0/api/java/org/apache/spark/sql/types/UserDefinedType.html)
+we need separate implementation for Python as well, so we need to provide the corresponding python class(s) in config params while
+creating the job or need to add them during runtime, see below for examples:
```
Comment (mayanks) [inline README.md:651]: Just wordsmithing here.
```
Spark's UserDefinedType needs a separate implementation for Python. Corresponding python class(s) should be provided as config params while creating the job or added during runtime. See examples below:
```

File: README.md [hunk 6]
```diff
@@ -619,6 +619,79 @@ the field's description which includes the spark type of the field.
...
 Row (i.e. not a field in a struct).
 
+#### BigNumeric support
+BigQuery's BigNumeric has a precision of 76.76 (the 77th digit is partial) and scale of 38. Since this much precision
+and scale is beyond the supported 38 scale and 38 precision of spark's DecimalType, the BigNumeric DataType is
+converted into spark's [UserDefinedType](https://spark.apache.org/docs/1.4.0/api/java/org/apache/spark/sql/types/UserDefinedType.html).
+The BigNumeric data can be accessed via BigNumericUDT DataType which internally uses [java.math.BigDecimal](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html) 
+to hold the BigNumeric data. The data can be read in either AVRO or ARROW formats.
+
+In order to write BigNumericUDT to BigQuery, use either ORC or PARQUET intermediate formats. Notice that the data gets
```
Comment (mayanks) [inline README.md:629]: Should we call out if there are any unsupported formats

File: README.md [hunk 7]
```diff
@@ -619,6 +619,79 @@ the field's description which includes the spark type of the field.
 In order to write those types to BigQuery, use the ORC or Avro intermediate format, and have them as column of the
 Row (i.e. not a field in a struct).
 
+#### BigNumeric support
+BigQuery's BigNumeric has a precision of 76.76 (the 77th digit is partial) and scale of 38. Since this much precision
```
Comment (mayanks) [inline README.md:623]:
```
BigQuery's BigNumeric has a precision of 76.76 (the 77th digit is partial) and scale of 38. This precision and scale is beyond spark's DecimalType(38 scale and 38 precision) support. The BigNumeric DataType is converted into spark's [UserDefinedType]
```

File: README.md (hunk 6 above)
Comment (davidrabinowitz) [reply README.md:629]: True, but we need to do it in another PR, not this one.

File: connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumericUDT.scala [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  override def userClass: Class[BigNumeric] = classOf[BigNumeric]
+
+  override def pyUDT: String =
+    "google.cloud.spark.bigquery.big_numeric_support.BigNumericUDT"
+}
+
+@SQLUserDefinedType(udt = classOf[BigNumericUDT])
+case class BigNumeric(number: BigDecimal){
```
Comment (davidrabinowitz) [inline connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumericUDT.scala:37]: `toString` needs to be overridden in order to reflect just the number, not `BigNumeric(1.1000...`

File: connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumericUDT.scala [hunk 9]
```diff
@@ -0,0 +1,38 @@
+package org.apache.spark.bigquery
```
Comment (davidrabinowitz) [inline connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumericUDT.scala:1]: Can you please port to Java, as we will need BigNumeric support in the DSv2 java only implementations as well

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
Where is the zip file created?

File: README.md (hunk 6 above)
Comment (himanshukohli09) [reply README.md]: I am explicitly mentioning (as a precautionary measure) that we do not support AVRO.

File: README.md (hunk 7 above)
Comment (himanshukohli09) [reply README.md]: Done, I have taken a portion from your suggestion, please have a look now.

**Comment by himanshukohli09** [comment]
> Where is the zip file created?

it gets created in the pythonlib folder

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
How the python distributable zip is created?

File: README.md [hunk 10]
```diff
@@ -636,49 +637,49 @@ Code examples:
...
 
 1) Adding python files while launching pyspark
 ```
-# spark-bigquery-latest_2.11.jar or spark-bigquery-latest_2.12.jar depending on scala version
-pyspark --jars gs://spark-lib/bigquery/spark-bigquery-latest_2.11.jar
-  --py-files gs://spark-lib/bigquery/bignumeric_support.zip
-  --files gs://spark-lib/bigquery/bignumeric_support.zip
+# use appropriate version for jar depending on the scala version
+pyspark --jars gs://spark-lib/bigquery/spark-bigquery-with-dependencies_2.11-0.21.0.jar
+  --py-files gs://spark-lib/bigquery/spark-bigquery-support_1.0.zip
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply README.md:658]: done

File: connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumericUDT.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        InternalRow row = new GenericInternalRow(1);
+        row.update(0, UTF8String.fromString(obj.getNumber().toPlainString()));
+        return row;
+    }
+
+    @Override
+    public BigNumeric deserialize(Object datum) {
+        if(!(datum instanceof UTF8String)) {
+            throw new IllegalArgumentException(
+                    "Not able to deserialize, was expecting an instance of UTF8String");
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumericUDT.java:30]: done

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: README.md [hunk 12]
```diff
@@ -619,6 +619,79 @@ the field's description which includes the spark type of the field.
 In order to write those types to BigQuery, use the ORC or Avro intermediate format, and have them as column of the
 Row (i.e. not a field in a struct).
 
+#### BigNumeric support
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply README.md]: I have moved the readme changes to a different PR: https://github.com/GoogleCloudDataproc/spark-bigquery-connector/pull/418

**Review by davidrabinowitz — APPROVED** [review]

**Comment by Skargonas** [comment]
I understand this change allows reading BigNumeric from BQ, but does it allow writing to BQ BigNumeric? In my experience, trying to write a Spark DecimalType with scale bigger than 9 would result in error, should I create a separate issue for that?

**Comment by himanshukohli09** [comment]
Hi Rafael,

This change allows reading BigNumeric from BQ, but while writing to BQ it writes in String format because of the reason mentioned by you "trying to write a Spark DecimalType with scale bigger than 9 would result in error". 

This way of supporting non spark Datatypes is similar to how we support DATETIME and TIME (https://github.com/GoogleCloudDataproc/spark-bigquery-connector#data-types).

**Comment by t-zaton** [comment]
Hi @himanshukohli09 ,
is the zip file for BigNumeric support still available somewhere in the GCS? It doesn't seem to exist in the path mentioned in the readme on the main page:
`gs://spark-lib/bigquery/spark-bigquery-support-0.22.0.zip`

**Pull request merged**